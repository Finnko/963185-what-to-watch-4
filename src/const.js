const Config = {
  MOVIE_MAX_RATING: 10,
  MOVIE_ACTORS_TO_SHOW: 4,
  SIMILAR_MOVIES_TO_SHOW: 4,
  MOVIES_NUMBER_TO_SHOW: 16,
  MOVIES_NUMBER_LIMIT: 8,
  MOVIES_NUMBER_OFFSET: 8,
  MAX_GENRES_TO_SHOW: 9,
  MOVIE_RATING_MAP: {
    0: `Bad`,
    3: `Normal`,
    5: `Good`,
    8: `Very good`,
    10: `Awesome`
  },
  ERRORS: {
    FETCH_DATA: `Something went wrong, please try again later.`,
  },
  LOADER: {
    SMALL: `small`,
    MEDIUM: `medium`,
  },
  DEFAULT_FILTER: `All genres`,
};

const PathName = {
  ROOT: `/`,
  MOVIE_PAGE: `/movie/`,
};

const TabName = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};

const ViewMode = {
  PROMO: {
    MAIN: `MAIN`,
    DETAILS: `DETAILS`
  },
  MOVIE_CARD: {
    WITH_PLAYER: `WITH_PLAYER`,
    IMAGE: `IMAGE`,
  }
};

export {
  Config,
  PathName,
  TabName,
  ViewMode,
};
