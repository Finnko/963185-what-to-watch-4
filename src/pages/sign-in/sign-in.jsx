import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Operation as UserOperation} from '../../store/reducers/user/operations';
import {getErrorStatus, getLoadingStatus} from '../../store/reducers/user/selectors';
import {Errors, LoaderSetup} from '../../const';
import Header from '../../components/header/header.jsx';
import Footer from '../../components/footer/footer.jsx';
import Input from '../../components/input/input.jsx';
import Loader from '../../components/loader/loader.tsx';

const inputNames = {
  email: `Email address`,
  password: `Password`,
};


class SignIn extends PureComponent {
  constructor(props) {
    super(props);

    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  handleFormSubmit(evt) {
    evt.preventDefault();

    const {email, password} = this.props.formControls;
    this.props.onFormSubmit({
      email: email.value,
      password: password.value,
    });
  }

  renderInputs() {
    const {onInputChange, formControls} = this.props;

    return Object.keys(formControls).map((name, index) => {
      const {
        value,
        valid,
        touched,
        validation
      } = formControls[name];

      return (
        <Input
          key={name + index}
          id={`user-${name}`}
          name={name}
          label={inputNames[name]}
          placeholder={inputNames[name]}
          value={value}
          valid={valid}
          touched={touched}
          shouldValidate={!!validation}
          onInputChange={onInputChange}
        />
      );
    });
  }

  render() {
    const {
      loading,
      error,
      formControls,
      isFormValid,
    } = this.props;
    const {email, password} = formControls;

    return (
      <div className="user-page">
        <Header className={`user-page__head`}/>

        <div className="sign-in user-page__content">
          <form
            action="#"
            className="sign-in__form"
            onSubmit={this.handleFormSubmit}
          >
            <div className="sign-in__message">
              {!email.valid && email.touched && <p>{`${Errors.WRONG_EMAIL}`}</p>}
              {!password.valid && password.touched && <p>{`${Errors.WRONG_PASSWORD}${password.validation.minLength}`}</p>}
              {error && <p>{`${Errors.WRONG_LOGIN_DATA}`}</p>}
            </div>

            <div className="sign-in__fields">
              {this.renderInputs()}
            </div>

            <div className="sign-in__submit">

              <button
                className="sign-in__btn"
                type="submit"
                disabled={!isFormValid}
              >
                {loading &&
                  <Loader
                    size={LoaderSetup.SIZE.SMALL}
                    position={LoaderSetup.POSITION.ABSOLUTE}
                  />
                }
                {!loading && `Sign in` }
              </button>

            </div>
          </form>
        </div>

        <Footer/>
      </div>
    );
  }
}

SignIn.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.bool.isRequired,
  formControls: PropTypes.object,
  isFormValid: PropTypes.bool.isRequired,
  onInputChange: PropTypes.func.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  return {
    loading: getLoadingStatus(state),
    error: getErrorStatus(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  onFormSubmit(authData) {
    dispatch(UserOperation.login(authData));
  }
});

export {SignIn};
export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
