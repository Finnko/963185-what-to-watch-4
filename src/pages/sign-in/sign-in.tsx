import * as React from 'react';
import {connect} from 'react-redux';
import {Operation as UserOperation} from '../../store/reducers/user/operations';
import {getErrorStatus, getLoadingStatus} from '../../store/reducers/user/selectors';
import {Errors, LoaderSetup} from '../../const';
import Header from '../../components/header/header';
import Footer from '../../components/footer/footer';
import Input from '../../components/input/input';
import Loader from '../../components/loader/loader';

const inputNames = {
  email: `Email address`,
  password: `Password`,
};

type SignInProps = {
  loading: boolean,
  error: boolean,
  formControls: any,
  isFormValid: boolean,
  onInputChange: () => void,
  onFormSubmit: ({}) => void,
}

class SignIn extends React.PureComponent<SignInProps> {
  constructor(props) {
    super(props);

    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  handleFormSubmit(evt: React.FormEvent<HTMLFormElement>) {
    evt.preventDefault();

    const {email, password} = this.props.formControls;
    this.props.onFormSubmit({
      email: email.value,
      password: password.value,
    });
  }

  renderInputs() {
    const {onInputChange, formControls, loading} = this.props;

    return Object.keys(formControls).map((name, index) => {
      const {
        value,
        valid,
        touched,
        validation
      } = formControls[name];

      return (
        <Input
          key={name + index}
          id={`user-${name}`}
          name={name}
          label={inputNames[name]}
          placeholder={inputNames[name]}
          value={value}
          valid={valid}
          touched={touched}
          shouldValidate={!!validation}
          disabled={loading}
          onInputChange={onInputChange}
        />
      );
    });
  }

  render() {
    const {
      loading,
      error,
      formControls,
      isFormValid,
    } = this.props;
    const {email, password} = formControls;

    return (
      <div className="user-page">
        <Header className={`user-page__head`}/>

        <div className="sign-in user-page__content">
          <form
            action="#"
            className="sign-in__form"
            onSubmit={this.handleFormSubmit}
          >
            <div className="sign-in__message">
              {!email.valid && email.touched && <p>{`${Errors.WRONG_EMAIL}`}</p>}
              {!password.valid && password.touched && <p>{`${Errors.WRONG_PASSWORD}${password.validation.minLength}`}</p>}
              {error && <p>{`${Errors.WRONG_LOGIN_DATA}`}</p>}
            </div>

            <div className="sign-in__fields">
              {this.renderInputs()}
            </div>

            <div className="sign-in__submit">

              <button
                className="sign-in__btn"
                type="submit"
                disabled={!isFormValid}
              >
                {loading &&
                  <Loader
                    size={LoaderSetup.SIZE.SMALL}
                    style={LoaderSetup.POSITION.ABSOLUTE}
                  />
                }
                {!loading && `Sign in` }
              </button>

            </div>
          </form>
        </div>

        <Footer/>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    loading: getLoadingStatus(state),
    error: getErrorStatus(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  onFormSubmit(authData) {
    dispatch(UserOperation.login(authData));
  }
});

export {SignIn};
export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
