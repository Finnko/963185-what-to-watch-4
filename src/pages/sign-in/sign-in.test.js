import React from 'react';
import renderer from 'react-test-renderer';
import {createMemoryHistory} from 'history';
import {Router} from 'react-router-dom';
import configureStore from 'redux-mock-store';
import NameSpace from '../../store/name-space';
import {Provider} from 'react-redux';
import {AuthStatus} from '../../const';
import {SignIn} from './sign-in.jsx';

const mockUser = {
  id: 1,
  email: `Oliver.conner@gmail.com`,
  name: `Oliver.conner`,
  avatarUrl: `img/1.png`
};

const mockStore = configureStore([]);

const store = mockStore({
  [NameSpace.USER]: {
    user: mockUser,
    loading: false,
    error: false,
    authStatus: AuthStatus.NO_AUTH,
  }
});

const props = {
  formControls: {
    email: {
      value: ``,
      valid: false,
      touched: false,
    },
    password: {
      value: ``,
      valid: false,
      touched: false,
    }
  },
  isFormValid: true,
  onFormSubmit: () => {},
  onInputChange: () => {},
};

describe(`SignIn component render correctly`, () => {
  const history = createMemoryHistory();

  it(`Should SignIn component render correctly`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <SignIn
                {...props}
                error={false}
                loading={false}
              />
            </Router>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should SignIn component render correctly with loading`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <SignIn
                {...props}
                error={false}
                loading={true}
              />
            </Router>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should SignIn component render correctly with error`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <SignIn
                {...props}
                error={true}
                loading={false}
              />
            </Router>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
