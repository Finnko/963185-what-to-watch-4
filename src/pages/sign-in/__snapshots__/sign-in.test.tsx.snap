// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn component render correctly Should SignIn component render correctly 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="user-block"
    >
      <a
        className="user-block__link"
        href="/login"
        onClick={[Function]}
      >
        Sign In
      </a>
    </div>
  </header>
  <div
    className="sign-in user-page__content"
  >
    <form
      action="#"
      className="sign-in__form"
      onSubmit={[Function]}
    >
      <div
        className="sign-in__message"
      />
      <div
        className="sign-in__fields"
      >
        <div
          className="sign-in__field "
        >
          <input
            className="sign-in__input"
            disabled={false}
            id="user-email"
            name="email"
            onChange={[Function]}
            placeholder="Email address"
            type="email"
            value=""
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-email"
          >
            Email address
          </label>
        </div>
        <div
          className="sign-in__field "
        >
          <input
            className="sign-in__input"
            disabled={false}
            id="user-password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-password"
          >
            Password
          </label>
        </div>
      </div>
      <div
        className="sign-in__submit"
      >
        <button
          className="sign-in__btn"
          disabled={false}
          type="submit"
        >
          Sign in
        </button>
      </div>
    </form>
  </div>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        © 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;

exports[`SignIn component render correctly Should SignIn component render correctly with error 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="user-block"
    >
      <a
        className="user-block__link"
        href="/login"
        onClick={[Function]}
      >
        Sign In
      </a>
    </div>
  </header>
  <div
    className="sign-in user-page__content"
  >
    <form
      action="#"
      className="sign-in__form"
      onSubmit={[Function]}
    >
      <div
        className="sign-in__message"
      >
        <p>
          We can’t recognize this email and password combination. Please try again.
        </p>
      </div>
      <div
        className="sign-in__fields"
      >
        <div
          className="sign-in__field "
        >
          <input
            className="sign-in__input"
            disabled={false}
            id="user-email"
            name="email"
            onChange={[Function]}
            placeholder="Email address"
            type="email"
            value=""
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-email"
          >
            Email address
          </label>
        </div>
        <div
          className="sign-in__field "
        >
          <input
            className="sign-in__input"
            disabled={false}
            id="user-password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-password"
          >
            Password
          </label>
        </div>
      </div>
      <div
        className="sign-in__submit"
      >
        <button
          className="sign-in__btn"
          disabled={false}
          type="submit"
        >
          Sign in
        </button>
      </div>
    </form>
  </div>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        © 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;

exports[`SignIn component render correctly Should SignIn component render correctly with loading 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="user-block"
    >
      <a
        className="user-block__link"
        href="/login"
        onClick={[Function]}
      >
        Sign In
      </a>
    </div>
  </header>
  <div
    className="sign-in user-page__content"
  >
    <form
      action="#"
      className="sign-in__form"
      onSubmit={[Function]}
    >
      <div
        className="sign-in__message"
      />
      <div
        className="sign-in__fields"
      >
        <div
          className="sign-in__field "
        >
          <input
            className="sign-in__input"
            disabled={true}
            id="user-email"
            name="email"
            onChange={[Function]}
            placeholder="Email address"
            type="email"
            value=""
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-email"
          >
            Email address
          </label>
        </div>
        <div
          className="sign-in__field "
        >
          <input
            className="sign-in__input"
            disabled={true}
            id="user-password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-password"
          >
            Password
          </label>
        </div>
      </div>
      <div
        className="sign-in__submit"
      >
        <button
          className="sign-in__btn"
          disabled={false}
          type="submit"
        >
          <div
            className="loader-wrapper"
            style={
              Object {
                "position": "absolute",
              }
            }
          >
            <div
              className="loader small"
            />
          </div>
        </button>
      </div>
    </form>
  </div>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        © 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
